modules/core/src/main/java/org/apache/ignite/internal/processors/query/aware/IndexBuildStatus.java

```diff
@@ -0,0 +1,168 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.aware;
+
+import org.apache.ignite.internal.util.typedef.internal.S;
+
+import static org.apache.ignite.internal.processors.query.aware.IndexBuildStatus.Status.COMPLETED;
+import static org.apache.ignite.internal.processors.query.aware.IndexBuildStatus.Status.DELETE;
+import static org.apache.ignite.internal.processors.query.aware.IndexBuildStatus.Status.INIT;
+
+/**
+ * Cache index build status.
+ * The following operations affect the status: rebuilding indexes and building a new index.
+ * <p/>
+ * If the operation starts executing, then the status is {@link Status#INIT INIT}.
+ * If all operations are completed, then the status is {@link Status#COMPLETED COMPLETED}.
+ * Status {@link Status#DELETE DELETE} is used for persistent cache to mark at the
+ * beginning of a checkpoint that all operations have been completed and they will be committed to it.
+ */
+public class IndexBuildStatus {
+    /** Enumeration of statuses. */
+    public enum Status {
+        /** Initial status - operation(s) in progress. */
+        INIT,
+
+        /** All operations completed. */
+        COMPLETED,
+
+        /** To be deleted. */
+        DELETE
+    }
+
+    /** Persistent cache. */
+    private final boolean persistent;
+
+    /** Status. Guarded by {@code this}. */
+    private Status status;
+
+    /** Rebuilding indexes. Guarded by {@code this}. */
+    private boolean rebuild;
+
+    /** Count of new indexes being built. Guarded by {@code this}. */
+    private int newIdx;
+
+    /**
+     * Constructor.
+     *
+     * @param persistent Persistent cache.
+     * @param rebuild {@code True} if rebuilding indexes, otherwise building a new index.
+     */
+    public IndexBuildStatus(boolean persistent, boolean rebuild) {
+        this.persistent = persistent;
+
+        onStartOperation(rebuild);
+    }
+
+    /**
+     * Callback on the start of of the operation.
+     *
+     * @param rebuild {@code True} if rebuilding indexes, otherwise building a new index.
+     * @see #onFinishOperation
+     */
+    public synchronized void onStartOperation(boolean rebuild) {
+        status = INIT;
+
+        if (rebuild)
+            this.rebuild = true;
+        else
+            newIdx++;
+    }
+
+    /**
+     * Callback on the finish of the operation.
+     *
+     * @param rebuild {@code True} if rebuilding indexes, otherwise building a new index.
+     * @return Current status.
+     * @see #onStartOperation
+     */
+    public synchronized Status onFinishOperation(boolean rebuild) {
+        if (rebuild)
+            this.rebuild = false;
+        else
+            newIdx = Math.max(0, newIdx - 1);
```

Comment:
If we have such logic (i mean Math.max) i suppose that we can`t rely on newIdx instance as exact, near i found a getter method which used only in tests, if we can`t make this logic more accurate for some reasons i suggest make getter as private (access by reflection) and append additional documentation.